From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 1 Mar 2025 16:49:39 +0100
Subject: [PATCH 081/225] WIP: remoteproc: add Unisoc remoteproc drivers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add drivers for booting remote processors found on the UMS9230 SoC.
This commit should be split up.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/remoteproc/Kconfig       |  27 ++
 drivers/remoteproc/Makefile      |   4 +
 drivers/remoteproc/sprd_common.c |  73 ++++
 drivers/remoteproc/sprd_common.h |  24 ++
 drivers/remoteproc/sprd_modem.c  | 357 +++++++++++++++++
 drivers/remoteproc/sprd_pmsys.c  | 229 +++++++++++
 drivers/remoteproc/sprd_wcn.c    | 653 +++++++++++++++++++++++++++++++
 7 files changed, 1367 insertions(+)
 create mode 100644 drivers/remoteproc/sprd_common.c
 create mode 100644 drivers/remoteproc/sprd_common.h
 create mode 100644 drivers/remoteproc/sprd_modem.c
 create mode 100644 drivers/remoteproc/sprd_pmsys.c
 create mode 100644 drivers/remoteproc/sprd_wcn.c

diff --git a/drivers/remoteproc/Kconfig b/drivers/remoteproc/Kconfig
index ee54436fe..bda2a7939 100644
--- a/drivers/remoteproc/Kconfig
+++ b/drivers/remoteproc/Kconfig
@@ -300,6 +300,33 @@ config RCAR_REMOTEPROC
 	  This can be either built-in or a loadable module.
 	  If compiled as module (M), the module name is rcar_rproc.
 
+config SPRD_REMOTEPROC_COMMON
+	tristate
+
+config SPRD_MODEM_REMOTEPROC
+	tristate "Unisoc integrated modem processor support"
+	depends on ARCH_SPRD || COMPILE_TEST
+	select SPRD_REMOTEPROC_COMMON
+	help
+	  Say y here to support the integrated modem processor (also
+	  known as PUBCP) found on Spreadtrum/Unisoc SoCs.
+
+config SPRD_PMSYS_REMOTEPROC
+	tristate "Unisoc PMSYS remoteproc support"
+	depends on ARCH_SPRD || COMPILE_TEST
+	select SPRD_REMOTEPROC_COMMON
+	help
+	  Say y here to support the PMSYS remote processor found on
+	  Spreadtrum/Unisoc SoCs.
+
+config SPRD_WCN_REMOTEPROC
+	tristate "Unisoc wireless connectivity processor support"
+	depends on ARCH_SPRD || COMPILE_TEST
+	select SPRD_REMOTEPROC_COMMON
+	help
+	  Say y here to support the integrated WCN processor found on
+	  Spreadtrum/Unisoc SoCs.
+
 config ST_REMOTEPROC
 	tristate "ST remoteproc support"
 	depends on ARCH_STI
diff --git a/drivers/remoteproc/Makefile b/drivers/remoteproc/Makefile
index 1c7598b84..d7a3c005e 100644
--- a/drivers/remoteproc/Makefile
+++ b/drivers/remoteproc/Makefile
@@ -33,6 +33,10 @@ obj-$(CONFIG_QCOM_WCNSS_PIL)		+= qcom_wcnss_pil.o
 qcom_wcnss_pil-y			+= qcom_wcnss.o
 qcom_wcnss_pil-y			+= qcom_wcnss_iris.o
 obj-$(CONFIG_RCAR_REMOTEPROC)		+= rcar_rproc.o
+obj-$(CONFIG_SPRD_REMOTEPROC_COMMON)	+= sprd_common.o
+obj-$(CONFIG_SPRD_MODEM_REMOTEPROC)	+= sprd_modem.o
+obj-$(CONFIG_SPRD_PMSYS_REMOTEPROC)	+= sprd_pmsys.o
+obj-$(CONFIG_SPRD_WCN_REMOTEPROC)	+= sprd_wcn.o
 obj-$(CONFIG_ST_REMOTEPROC)		+= st_remoteproc.o
 obj-$(CONFIG_ST_SLIM_REMOTEPROC)	+= st_slim_rproc.o
 obj-$(CONFIG_STM32_RPROC)		+= stm32_rproc.o
diff --git a/drivers/remoteproc/sprd_common.c b/drivers/remoteproc/sprd_common.c
new file mode 100644
index 000000000..17606149f
--- /dev/null
+++ b/drivers/remoteproc/sprd_common.c
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Unisoc remoteproc helpers
+ *
+ * Copyright (C) 2024 Otto Pflüger
+ */
+
+#include <linux/rpmsg/sprd_sipc.h>
+
+#include "sprd_common.h"
+
+#define to_sipc_subdev(_sd) container_of(_sd, struct sprd_sipc_subdev, subdev)
+
+static int sipc_subdev_prepare(struct rproc_subdev *rpsubdev)
+{
+	struct sprd_sipc_subdev *sd = to_sipc_subdev(rpsubdev);
+
+	sd->sipc = sprd_sipc_register(sd->dev, sd->node);
+
+	return PTR_ERR_OR_ZERO(sd->sipc);
+}
+
+static void sipc_subdev_stop(struct rproc_subdev *rpsubdev, bool crashed)
+{
+	struct sprd_sipc_subdev *sd = to_sipc_subdev(rpsubdev);
+
+	sprd_sipc_unregister(sd->sipc);
+	sd->sipc = NULL;
+}
+
+static void sipc_subdev_unprepare(struct rproc_subdev *rpsubdev)
+{
+	struct sprd_sipc_subdev *sd = to_sipc_subdev(rpsubdev);
+
+	/* unregister if stop wasn't called due to a start failure */
+	if (sd->sipc) {
+		sprd_sipc_unregister(sd->sipc);
+		sd->sipc = NULL;
+	}
+}
+
+static void remove_sipc_subdev(void *data)
+{
+	struct sprd_sipc_subdev *sd = data;
+
+	rproc_remove_subdev(sd->rproc, &sd->subdev);
+	of_node_put(sd->node);
+}
+
+int devm_sprd_rproc_add_sipc_subdev(struct rproc *rproc,
+				    struct device_node *np,
+				    struct sprd_sipc_subdev *sd)
+{
+	struct device *dev = &rproc->dev;
+
+	sd->node = of_get_child_by_name(np, "sipc");
+	if (!sd->node)
+		return 0;
+
+	sd->dev = dev;
+	sd->rproc = rproc;
+	sd->subdev.prepare = sipc_subdev_prepare;
+	sd->subdev.unprepare = sipc_subdev_unprepare;
+	sd->subdev.stop = sipc_subdev_stop;
+
+	rproc_add_subdev(rproc, &sd->subdev);
+
+	return devm_add_action_or_reset(rproc->dev.parent, remove_sipc_subdev, sd);
+}
+EXPORT_SYMBOL_GPL(devm_sprd_rproc_add_sipc_subdev);
+
+MODULE_DESCRIPTION("Unisoc remoteproc helper driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/remoteproc/sprd_common.h b/drivers/remoteproc/sprd_common.h
new file mode 100644
index 000000000..26eec6394
--- /dev/null
+++ b/drivers/remoteproc/sprd_common.h
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0
+#ifndef __RPROC_SPRD_COMMON_H__
+#define __RPROC_SPRD_COMMON_H__
+
+#include <linux/remoteproc.h>
+
+#include "remoteproc_internal.h"
+
+struct sprd_sipc;
+
+struct sprd_sipc_subdev {
+	struct rproc_subdev subdev;
+
+	struct rproc *rproc;
+	struct device *dev;
+	struct device_node *node;
+	struct sprd_sipc *sipc;
+};
+
+int devm_sprd_rproc_add_sipc_subdev(struct rproc *rproc,
+				    struct device_node *np,
+				    struct sprd_sipc_subdev *subdev);
+
+#endif
diff --git a/drivers/remoteproc/sprd_modem.c b/drivers/remoteproc/sprd_modem.c
new file mode 100644
index 000000000..55689adf2
--- /dev/null
+++ b/drivers/remoteproc/sprd_modem.c
@@ -0,0 +1,357 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 Otto Pflüger
+ */
+
+#include <linux/mfd/syscon.h>
+#include <linux/module.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
+#include <linux/regmap.h>
+#include <linux/remoteproc.h>
+#include <linux/reset.h>
+
+#include "sprd_common.h"
+
+struct sprd_modem_region_header {
+	char name[24];
+	__le64 base;
+	__le64 size;
+	__le64 flags;
+} __packed;
+
+struct sprd_modem_image_header {
+	__be32 magic;
+	__be32 version;
+	__le32 length;
+	__le32 unused;
+	u8 rsvmem_info[4 * 48];
+	__le32 bootcode_len;
+	u8 bootcode[200];
+	u32 unused2;
+	struct sprd_modem_region_header region_info[8];
+	/* more fields exist, but are not used here */
+} __packed;
+
+struct sprd_modem_info {
+	u32 corereset_reg;
+	u32 corereset_mask;
+};
+
+struct sprd_modem {
+	struct device *dev;
+	struct sprd_sipc_subdev sipc;
+	struct reset_control *reset;
+	struct regmap *aon_apb_regs;
+	const struct sprd_modem_info *info;
+	phys_addr_t mem_base;
+	size_t mem_size;
+	void *bootmem;
+};
+
+struct sprd_modem_region_def {
+	const char *name;
+	const char *filename;
+	bool zero;
+};
+
+const struct sprd_modem_region_def sprd_modem_region_defs[] = {
+	{ .name = "modem", .filename = NULL /* use main image */ },
+	{ .name = "cpcmdline", .zero = true },
+	{ .name = "deltanv", .filename = "sprd/l_deltanv.bin" },
+	{ .name = "fixnv", .filename = "sprd/l_fixnv.bin" },
+	{ .name = "runnv", .filename = "sprd/l_runtimenv.bin" },
+	{ .name = "gdsp", .filename = "sprd/l_gdsp.bin" },
+	{ .name = "ldsp", .filename = "sprd/l_ldsp.bin" },
+	{ /* sentinel */ }
+};
+
+static int sprd_modem_load_region(struct sprd_modem *m,
+				  const struct firmware *fw,
+				  void *mem, int idx,
+				  const struct sprd_modem_region_def *rdef,
+				  struct sprd_modem_region_header *rhdr)
+{
+	u64 region_start, region_end, region_size;
+	int ret;
+
+	if (rdef->filename)
+		dev_dbg(m->dev, "loading file %s into region %s\n", rdef->filename, rdef->name);
+	else
+		dev_dbg(m->dev, "loading base image into region %s\n", rdef->name);
+
+	region_start = le64_to_cpu(rhdr->base);
+	region_size = le64_to_cpu(rhdr->size);
+	region_end = region_start + region_size;
+
+	if (region_end > (m->mem_base + m->mem_size) || region_start < m->mem_base) {
+		dev_err(m->dev, "region %d (%s) outside memory range\n", idx, rdef->name);
+		return -EINVAL;
+	}
+
+	if (rdef->filename) {
+		ret = request_firmware(&fw, rdef->filename, m->dev);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (rdef->zero) {
+		memset(mem + (region_start - m->mem_base), 0, region_size);
+	} else if (fw->size > region_size) {
+		if (rdef->filename)
+			release_firmware(fw);
+		dev_err(m->dev, "firmware too large for region %s\n", rdef->name);
+		return -ENOMEM;
+	} else {
+		memcpy(mem + (region_start - m->mem_base), fw->data, fw->size);
+	}
+
+	if (rdef->filename)
+		release_firmware(fw);
+
+	return 0;
+}
+
+static int sprd_modem_load(struct rproc *rproc, const struct firmware *fw)
+{
+	struct sprd_modem_image_header *hdr = (void *)fw->data;
+	const struct sprd_modem_region_def *rdef;
+	struct sprd_modem_region_header *rhdr;
+	struct sprd_modem *m = rproc->priv;
+	int i, ret = 0;
+	void *mem;
+
+	if (le32_to_cpu(hdr->bootcode_len) * 4 > sizeof(hdr->bootcode)) {
+		dev_err(m->dev, "invalid bootcode size in image\n");
+		return -EINVAL;
+	}
+
+	memcpy_toio(m->bootmem, hdr->bootcode, le32_to_cpu(hdr->bootcode_len) * 4);
+
+	mem = memremap(m->mem_base, m->mem_size, MEMREMAP_WC);
+	if (!mem) {
+		dev_err(m->dev, "failed to map memory region\n");
+		return -EBUSY;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(hdr->region_info); i++) {
+		rhdr = &hdr->region_info[i];
+		if (!rhdr->base)
+			break;
+
+		dev_dbg(m->dev, "found region %-*s at %08llx\n",
+			(int)sizeof(rhdr->name), rhdr->name, le64_to_cpu(rhdr->base));
+
+		for (rdef = sprd_modem_region_defs; rdef->name; rdef++) {
+			if (!strncmp(rhdr->name, rdef->name, sizeof(rhdr->name))) {
+				ret = sprd_modem_load_region(m, fw, mem, i, rdef, rhdr);
+				if (ret < 0)
+					goto out;
+			}
+		}
+	}
+
+out:
+	memunmap(mem);
+	return ret;
+}
+
+static int sprd_modem_start(struct rproc *rproc)
+{
+	struct sprd_modem *m = rproc->priv;
+	int ret;
+
+	pm_runtime_enable(m->dev);
+	ret = pm_runtime_resume_and_get(m->dev);
+	if (ret < 0)
+		goto err_pm_disable;
+
+	ret = reset_control_deassert(m->reset);
+	if (ret < 0)
+		goto err_pm_put;
+
+	/* start processor */
+	regmap_clear_bits(m->aon_apb_regs, m->info->corereset_reg,
+			  m->info->corereset_mask);
+
+	return 0;
+
+err_pm_put:
+	pm_runtime_put_sync(m->dev);
+err_pm_disable:
+	pm_runtime_disable(m->dev);
+	return ret;
+}
+
+static int sprd_modem_stop(struct rproc *rproc)
+{
+	struct sprd_modem *m = rproc->priv;
+
+	/* stop processor */
+	regmap_set_bits(m->aon_apb_regs, m->info->corereset_reg,
+			m->info->corereset_mask);
+
+	reset_control_assert(m->reset);
+
+	pm_runtime_put_sync_suspend(m->dev);
+	pm_runtime_disable(m->dev);
+
+	return 0;
+}
+
+static const struct rproc_ops sprd_modem_ops = {
+	.load		= sprd_modem_load,
+	.start		= sprd_modem_start,
+	.stop		= sprd_modem_stop,
+};
+
+static int sprd_modem_parse_memory_region(struct sprd_modem *m)
+{
+	struct device_node *np;
+	struct reserved_mem *rmem;
+
+	np = of_parse_phandle(m->dev->of_node, "memory-region", 0);
+	if (!np) {
+		dev_err(m->dev, "no memory region specified\n");
+		return -EINVAL;
+	}
+
+	rmem = of_reserved_mem_lookup(np);
+	if (!rmem) {
+		of_node_put(np);
+		dev_err(m->dev, "failed to look up memory region\n");
+		return -EINVAL;
+	}
+
+	m->mem_base = rmem->base;
+	m->mem_size = rmem->size;
+
+	of_node_put(np);
+
+	return 0;
+}
+
+static int sprd_modem_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct sprd_modem *m;
+	struct rproc *rproc;
+	const char *fw_name;
+	int ret;
+
+	ret = of_property_read_string_index(dev->of_node, "firmware-name",
+					    0, &fw_name);
+	if (ret < 0) {
+		dev_err(&pdev->dev, "unable to read firmware-name\n");
+		return ret;
+	}
+
+	rproc = devm_rproc_alloc(dev, dev->of_node->name, &sprd_modem_ops,
+				 fw_name, sizeof(*m));
+	if (!rproc)
+		return -ENOMEM;
+
+	platform_set_drvdata(pdev, rproc);
+
+	m = rproc->priv;
+	m->dev = dev;
+
+	m->info = of_device_get_match_data(dev);
+	if (!m->info)
+		return -EINVAL;
+
+	m->bootmem = devm_platform_get_and_ioremap_resource(pdev, 0, NULL);
+	if (IS_ERR(m->bootmem))
+		return PTR_ERR(m->bootmem);
+
+	ret = sprd_modem_parse_memory_region(m);
+	if (ret < 0)
+		return ret;
+
+	m->aon_apb_regs = syscon_regmap_lookup_by_phandle(dev->of_node, "sprd,syscon-aon-apb");
+	if (IS_ERR(m->aon_apb_regs)) {
+		dev_err(m->dev, "failed to get aon-apb syscon handle\n");
+		return PTR_ERR(m->aon_apb_regs);
+	}
+
+	m->reset = devm_reset_control_get(dev, NULL);
+	if (IS_ERR(m->reset)) {
+		dev_err(m->dev, "failed to get modem reset\n");
+		return PTR_ERR(m->reset);
+	}
+
+	device_set_wakeup_capable(dev, true);
+
+	if (of_property_read_bool(dev->of_node, "wakeup-source"))
+		device_wakeup_enable(dev);
+
+	ret = devm_sprd_rproc_add_sipc_subdev(rproc, dev->of_node, &m->sipc);
+	if (ret < 0) {
+		dev_err(dev, "failed to add SIPC subdev\n");
+		return ret;
+	}
+
+	ret = devm_rproc_add(dev, rproc);
+	if (ret < 0) {
+		dev_err(dev, "failed to add rproc\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static const struct sprd_modem_info ums9230_modem_info = {
+	.corereset_reg = 0x0174,
+	.corereset_mask = BIT(10),
+};
+
+static const struct of_device_id sprd_modem_of_match[] = {
+	{ .compatible = "sprd,ums9230-modem", .data = &ums9230_modem_info },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, sprd_modem_of_match);
+
+static int sprd_modem_suspend(struct device *dev)
+{
+	struct rproc *rproc = dev_get_drvdata(dev);
+
+	if (device_may_wakeup(dev))
+		return 0;
+
+	if (rproc->state == RPROC_RUNNING)
+		return sprd_modem_stop(rproc->priv);
+
+	return 0;
+}
+
+static int sprd_modem_resume(struct device *dev)
+{
+	struct rproc *rproc = dev_get_drvdata(dev);
+
+	if (device_may_wakeup(dev))
+		return 0;
+
+	if (rproc->state == RPROC_RUNNING)
+		return sprd_modem_start(rproc->priv);
+
+	return 0;
+}
+
+static struct dev_pm_ops sprd_modem_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(sprd_modem_suspend, sprd_modem_resume)
+};
+
+static struct platform_driver sprd_modem_driver = {
+	.probe = sprd_modem_probe,
+	.driver = {
+		.name = "sprd-modem",
+		.of_match_table = sprd_modem_of_match,
+		.pm = &sprd_modem_pm_ops,
+	},
+};
+
+module_platform_driver(sprd_modem_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_DESCRIPTION("Unisoc modem remoteproc driver");
diff --git a/drivers/remoteproc/sprd_pmsys.c b/drivers/remoteproc/sprd_pmsys.c
new file mode 100644
index 000000000..9685684d2
--- /dev/null
+++ b/drivers/remoteproc/sprd_pmsys.c
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 Otto Pflüger
+ */
+
+#include <linux/mfd/syscon.h>
+#include <linux/module.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/platform_device.h>
+#include <linux/regmap.h>
+#include <linux/remoteproc.h>
+#include <linux/reset.h>
+
+#include "sprd_common.h"
+
+struct sprd_pmsys_info {
+	u32 corereset_reg;
+	u32 corereset_mask;
+};
+
+struct sprd_pmsys {
+	struct device *dev;
+	struct sprd_sipc_subdev sipc;
+	struct reset_control *reset;
+	struct regmap *aon_apb_regs;
+	const struct sprd_pmsys_info *info;
+	phys_addr_t mem_base;
+	size_t mem_size;
+	const char *sys_fw_name;
+	void *bootmem;
+	size_t bootmem_size;
+};
+
+static int sprd_pmsys_load(struct rproc *rproc, const struct firmware *fw)
+{
+	struct sprd_pmsys *p = rproc->priv;
+	int ret = 0;
+	void *mem;
+
+	if (fw->size > p->bootmem_size) {
+		dev_err(p->dev, "bootcode firmware too large\n");
+		return -ENOMEM;
+	}
+
+	memcpy_toio(p->bootmem, fw->data, fw->size);
+
+	mem = memremap(p->mem_base, p->mem_size, MEMREMAP_WC);
+	if (!mem) {
+		dev_err(p->dev, "failed to map memory region\n");
+		return -EBUSY;
+	}
+
+	ret = request_firmware(&fw, p->sys_fw_name, p->dev);
+	if (ret)
+		goto out_unmap;
+
+	if (fw->size > p->mem_size) {
+		dev_err(p->dev, "pmsys firmware too large\n");
+		ret = -ENOMEM;
+		goto out_release;
+	}
+
+	memcpy(mem, fw->data, fw->size);
+
+out_release:
+	release_firmware(fw);
+out_unmap:
+	memunmap(mem);
+	return ret;
+}
+
+static int sprd_pmsys_start(struct rproc *rproc)
+{
+	struct sprd_pmsys *p = rproc->priv;
+	int ret;
+
+	ret = reset_control_deassert(p->reset);
+	if (ret < 0)
+		return ret;
+
+	/* start processor */
+	regmap_clear_bits(p->aon_apb_regs, p->info->corereset_reg,
+			  p->info->corereset_mask);
+
+	return 0;
+}
+
+static int sprd_pmsys_stop(struct rproc *rproc)
+{
+	struct sprd_pmsys *p = rproc->priv;
+
+	/* stop processor */
+	regmap_set_bits(p->aon_apb_regs, p->info->corereset_reg,
+			p->info->corereset_mask);
+
+	reset_control_assert(p->reset);
+
+	return 0;
+}
+
+static const struct rproc_ops sprd_pmsys_ops = {
+	.load		= sprd_pmsys_load,
+	.start		= sprd_pmsys_start,
+	.stop		= sprd_pmsys_stop,
+};
+
+static int sprd_pmsys_parse_memory_region(struct sprd_pmsys *p)
+{
+	struct device_node *np;
+	struct reserved_mem *rmem;
+
+	np = of_parse_phandle(p->dev->of_node, "memory-region", 0);
+	if (!np) {
+		dev_err(p->dev, "no memory region specified\n");
+		return -EINVAL;
+	}
+
+	rmem = of_reserved_mem_lookup(np);
+	if (!rmem) {
+		of_node_put(np);
+		dev_err(p->dev, "failed to look up memory region\n");
+		return -EINVAL;
+	}
+
+	p->mem_base = rmem->base;
+	p->mem_size = rmem->size;
+
+	of_node_put(np);
+
+	return 0;
+}
+
+static int sprd_pmsys_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct resource *bootmem;
+	struct sprd_pmsys *p;
+	struct rproc *rproc;
+	const char *fw_name;
+	int ret;
+
+	ret = of_property_read_string_index(dev->of_node, "firmware-name",
+					    0, &fw_name);
+	if (ret < 0) {
+		dev_err(&pdev->dev, "unable to read firmware-name\n");
+		return ret;
+	}
+
+	rproc = devm_rproc_alloc(dev, dev->of_node->name, &sprd_pmsys_ops,
+				 fw_name, sizeof(*p));
+	if (!rproc)
+		return -ENOMEM;
+
+	platform_set_drvdata(pdev, rproc);
+
+	p = rproc->priv;
+	p->dev = dev;
+
+	ret = of_property_read_string_index(dev->of_node, "firmware-name",
+					    1, &p->sys_fw_name);
+	if (ret < 0) {
+		dev_err(p->dev, "unable to read second firmware-name\n");
+		return ret;
+	}
+
+	p->info = of_device_get_match_data(dev);
+	if (!p->info)
+		return -EINVAL;
+
+	p->bootmem = devm_platform_get_and_ioremap_resource(pdev, 0, &bootmem);
+	if (IS_ERR(p->bootmem))
+		return PTR_ERR(p->bootmem);
+
+	p->bootmem_size = bootmem->end - bootmem->start + 1;
+
+	ret = sprd_pmsys_parse_memory_region(p);
+	if (ret < 0)
+		return ret;
+
+	p->aon_apb_regs = syscon_regmap_lookup_by_phandle(dev->of_node, "sprd,syscon-aon-apb");
+	if (IS_ERR(p->aon_apb_regs)) {
+		dev_err(p->dev, "failed to get aon-apb syscon handle\n");
+		return PTR_ERR(p->aon_apb_regs);
+	}
+
+	p->reset = devm_reset_control_get_optional(dev, NULL);
+	if (IS_ERR(p->reset)) {
+		dev_err(p->dev, "failed to get pmsys reset\n");
+		return PTR_ERR(p->reset);
+	}
+
+	ret = devm_sprd_rproc_add_sipc_subdev(rproc, dev->of_node, &p->sipc);
+	if (ret < 0) {
+		dev_err(dev, "failed to add SIPC subdev\n");
+		return ret;
+	}
+
+	ret = devm_rproc_add(dev, rproc);
+	if (ret < 0) {
+		dev_err(dev, "failed to add rproc\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static const struct sprd_pmsys_info ums9230_pmsys_info = {
+	.corereset_reg = 0x008c,
+	.corereset_mask = BIT(0),
+};
+
+static const struct of_device_id sprd_pmsys_of_match[] = {
+	{ .compatible = "sprd,ums9230-pmsys", .data = &ums9230_pmsys_info },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, sprd_pmsys_of_match);
+
+static struct platform_driver sprd_pmsys_driver = {
+	.probe = sprd_pmsys_probe,
+	.driver = {
+		.name = "sprd-pmsys",
+		.of_match_table = sprd_pmsys_of_match,
+	},
+};
+
+module_platform_driver(sprd_pmsys_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_DESCRIPTION("Unisoc PMSYS remoteproc driver");
diff --git a/drivers/remoteproc/sprd_wcn.c b/drivers/remoteproc/sprd_wcn.c
new file mode 100644
index 000000000..dd6168f48
--- /dev/null
+++ b/drivers/remoteproc/sprd_wcn.c
@@ -0,0 +1,653 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 Otto Pflüger
+ */
+
+#include <linux/gpio/consumer.h>
+#include <linux/mfd/syscon.h>
+#include <linux/module.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
+#include <linux/regmap.h>
+#include <linux/regulator/consumer.h>
+#include <linux/remoteproc.h>
+#include <linux/reset.h>
+
+#include "sprd_common.h"
+
+struct sprd_wcn_subsys_info {
+	u32 init_status_offset;
+	u32 init_status_boot_magic;
+	u32 sys_rst_reg;
+	u32 sys_rst_mask;
+	u32 sync_addr_reg;
+	u32 deep_sleep_reg;
+	u32 deep_sleep_mask;
+	u32 boot_sel_reg;
+};
+
+struct sprd_wcn_info {
+	const struct regulator_bulk_data *regulators;
+	unsigned int num_regulators;
+	u32 dram_offset;
+	u32 mem_remap_reg;
+	u32 mem_remap_val;
+	u32 ap_wcn_sleep_ctrl_reg;
+	u32 ap_wcn_sleep_allow_mask;
+	u32 deep_sleep_ctrl_reg;
+	u32 deep_sleep_ctrl_mask;
+	struct sprd_wcn_subsys_info btwf;
+	struct sprd_wcn_subsys_info gnss;
+};
+
+struct sprd_wcn_shared {
+	struct device *dev;
+	struct device btwf_dev;
+	struct device gnss_dev;
+
+	const struct sprd_wcn_info *info;
+	struct regulator_bulk_data *regulators;
+	struct gpio_desc *en_gpio;
+	struct gpio_desc *rst_gpio;
+	struct reset_control *reset;
+	struct regmap *ap_pub_apb_regs;
+	struct regmap *ap_aon_apb_regs;
+	struct regmap *wcn_aon_apb_regs;
+	struct regmap *wcn_aon_ahb_regs;
+
+	unsigned int rf_power_count;
+	unsigned int sleep_disable_count;
+
+	struct mutex lock;
+};
+
+struct sprd_wcn_proc {
+	struct device *dev;
+	struct sprd_wcn_shared *wcn;
+	struct sprd_sipc_subdev sipc;
+
+	const struct sprd_wcn_subsys_info *info;
+	struct regmap *wcn_ahb_regs;
+
+	phys_addr_t mem_base;
+	size_t mem_size;
+
+	phys_addr_t sync_base;
+	void __iomem *sync_virt;
+	size_t sync_size;
+};
+
+static int sprd_wcn_rf_power_on(struct sprd_wcn_shared *wcn)
+{
+	int ret;
+
+	if (wcn->rf_power_count++ != 0)
+		return 0;
+
+	ret = regulator_bulk_enable(wcn->info->num_regulators, wcn->regulators);
+	if (ret < 0) {
+		wcn->rf_power_count--;
+		return ret;
+	}
+
+	gpiod_set_value(wcn->en_gpio, 1);
+	usleep_range(500, 1000);
+	gpiod_set_value(wcn->rst_gpio, 0);
+
+	return 0;
+}
+
+static void sprd_wcn_rf_power_off(struct sprd_wcn_shared *wcn)
+{
+	if (WARN_ON(wcn->rf_power_count == 0))
+		return;
+
+	if (wcn->rf_power_count-- != 1)
+		return;
+
+	gpiod_set_value(wcn->rst_gpio, 1);
+	gpiod_set_value(wcn->en_gpio, 0);
+
+	regulator_bulk_disable(wcn->info->num_regulators, wcn->regulators);
+}
+
+static void sprd_wcn_sleep_disable(struct sprd_wcn_shared *wcn)
+{
+	if (wcn->sleep_disable_count++ != 0)
+		return;
+
+	regmap_write(wcn->ap_aon_apb_regs, wcn->info->ap_wcn_sleep_ctrl_reg + 0x2000,
+		     wcn->info->ap_wcn_sleep_allow_mask);
+}
+
+static void sprd_wcn_sleep_enable(struct sprd_wcn_shared *wcn)
+{
+	if (WARN_ON(wcn->sleep_disable_count == 0))
+		return;
+
+	if (wcn->sleep_disable_count-- != 1)
+		return;
+
+	regmap_write(wcn->ap_aon_apb_regs, wcn->info->ap_wcn_sleep_ctrl_reg + 0x1000,
+		     wcn->info->ap_wcn_sleep_allow_mask);
+}
+
+static int sprd_wcn_load(struct rproc *rproc, const struct firmware *fw)
+{
+	struct sprd_wcn_proc *wp = rproc->priv;
+	struct sprd_wcn_shared *wcn = wp->wcn;
+	void *mem;
+
+	if (fw->size > wp->mem_size) {
+		dev_err(wcn->dev, "firmware too large\n");
+		return -EINVAL;
+	}
+
+	mem = memremap(wp->mem_base, wp->mem_size, MEMREMAP_WC);
+	if (!mem) {
+		dev_err(wp->dev, "failed to map memory region\n");
+		return -EBUSY;
+	}
+	memcpy(mem, fw->data, fw->size);
+	memunmap(mem);
+
+	return 0;
+}
+
+static int sprd_wcn_poll_boot_status(struct sprd_wcn_proc *wp)
+{
+	unsigned int retries = 100;
+	u32 init_status;
+	int ret = -ETIMEDOUT;
+
+	while (retries--) {
+		init_status = readl(wp->sync_virt + wp->info->init_status_offset);
+		dev_dbg(wp->dev, "init status: %08x\n", init_status);
+		if (init_status == wp->info->init_status_boot_magic) {
+			ret = 0;
+			break;
+		}
+		msleep(20);
+	}
+
+	return ret;
+}
+
+static int sprd_wcn_start(struct rproc *rproc)
+{
+	struct sprd_wcn_proc *wp = rproc->priv;
+	struct sprd_wcn_shared *wcn = wp->wcn;
+	int ret;
+
+	mutex_lock(&wcn->lock);
+
+	ret = sprd_wcn_rf_power_on(wcn);
+	if (ret < 0)
+		goto err_out;
+
+	regmap_write(wcn->ap_pub_apb_regs, wcn->info->mem_remap_reg,
+		     wcn->info->mem_remap_val);
+
+	memset_io(wp->sync_virt, 0, wp->sync_size);
+
+	ret = pm_runtime_resume_and_get(wcn->dev);
+	if (ret < 0)
+		goto err_rf_off;
+
+	ret = reset_control_deassert(wcn->reset);
+	if (ret < 0)
+		goto err_pm_put;
+
+	sprd_wcn_sleep_disable(wcn);
+
+	/* wait for stable WCN power */
+	msleep(20);
+
+	/* assert cpu reset to avoid accidental startup */
+	regmap_set_bits(wcn->wcn_aon_ahb_regs, wp->info->sys_rst_reg,
+			wp->info->sys_rst_mask);
+
+	/* disable force deep sleep */
+	regmap_clear_bits(wcn->wcn_aon_apb_regs, wp->info->deep_sleep_reg,
+			  wp->info->deep_sleep_mask);
+	/* allow wcn to control deep sleep */
+	regmap_set_bits(wcn->wcn_aon_apb_regs, wcn->info->deep_sleep_ctrl_reg,
+			wcn->info->deep_sleep_ctrl_mask);
+	/* wait for stable subsystem power */
+	msleep(20);
+	/* select boot from dram */
+	regmap_update_bits(wp->wcn_ahb_regs, wp->info->boot_sel_reg,
+			   0x3ffffff, 0x3000000 |
+			   (u32)(wp->mem_base - wcn->info->dram_offset));
+	/* write shared memory address */
+	regmap_write(wcn->wcn_aon_ahb_regs, wp->info->sync_addr_reg,
+		     (u32)(wp->sync_base - wp->mem_base));
+	/* deassert cpu reset */
+	regmap_clear_bits(wcn->wcn_aon_ahb_regs, wp->info->sys_rst_reg,
+			  wp->info->sys_rst_mask);
+
+	sprd_wcn_sleep_enable(wcn);
+
+	ret = sprd_wcn_poll_boot_status(wp);
+	if (ret < 0)
+		goto err_cpu_stop;
+
+	mutex_unlock(&wcn->lock);
+
+	return 0;
+
+err_cpu_stop:
+	/* assert cpu reset */
+	regmap_set_bits(wcn->wcn_aon_ahb_regs, wp->info->sys_rst_reg,
+			wp->info->sys_rst_mask);
+	/* force deep sleep */
+	regmap_set_bits(wcn->wcn_aon_apb_regs, wp->info->deep_sleep_reg,
+			wp->info->deep_sleep_mask);
+	reset_control_assert(wcn->reset);
+err_pm_put:
+	pm_runtime_put_sync(wcn->dev);
+err_rf_off:
+	sprd_wcn_rf_power_off(wcn);
+err_out:
+	mutex_unlock(&wcn->lock);
+	return ret;
+}
+
+static int sprd_wcn_stop(struct rproc *rproc)
+{
+	struct sprd_wcn_proc *wp = rproc->priv;
+	struct sprd_wcn_shared *wcn = wp->wcn;
+
+	mutex_lock(&wcn->lock);
+
+	/* needed to access registers from AP */
+	sprd_wcn_sleep_disable(wcn);
+	/* wait for stable power */
+	msleep(40);
+
+	/* assert reset */
+	regmap_set_bits(wcn->wcn_aon_ahb_regs, wp->info->sys_rst_reg,
+			wp->info->sys_rst_mask);
+	/* force deep sleep */
+	regmap_set_bits(wcn->wcn_aon_apb_regs, wp->info->deep_sleep_reg,
+			wp->info->deep_sleep_mask);
+
+	sprd_wcn_sleep_enable(wcn);
+	reset_control_assert(wcn->reset);
+	pm_runtime_put_sync_suspend(wcn->dev);
+	sprd_wcn_rf_power_off(wcn);
+
+	mutex_unlock(&wcn->lock);
+
+	return 0;
+}
+
+static const struct rproc_ops wcn_ops = {
+	.load		= sprd_wcn_load,
+	.start		= sprd_wcn_start,
+	.stop		= sprd_wcn_stop,
+};
+
+static int sprd_wcn_parse_memory_region(struct sprd_wcn_proc *wp)
+{
+	struct device_node *np;
+	struct reserved_mem *rmem;
+
+	np = of_parse_phandle(wp->dev->of_node, "memory-region", 0);
+	if (!np) {
+		dev_err(wp->dev, "no memory region specified for\n");
+		return -EINVAL;
+	}
+
+	rmem = of_reserved_mem_lookup(np);
+	if (!rmem) {
+		of_node_put(np);
+		dev_err(wp->dev, "failed to look up memory region\n");
+		return -EINVAL;
+	}
+
+	wp->mem_base = rmem->base;
+	wp->mem_size = rmem->size;
+
+	of_node_put(np);
+
+	np = of_parse_phandle(wp->dev->of_node, "sprd,sync-region", 0);
+	if (!np) {
+		dev_err(wp->dev, "no sync region specified\n");
+		return -EINVAL;
+	}
+
+	rmem = of_reserved_mem_lookup(np);
+	if (!rmem) {
+		of_node_put(np);
+		dev_err(wp->dev, "failed to look up sync region\n");
+		return -EINVAL;
+	}
+
+	wp->sync_base = rmem->base;
+	wp->sync_size = rmem->size;
+	wp->sync_virt = devm_ioremap_wc(wp->dev, rmem->base, rmem->size);
+	if (!wp->sync_virt) {
+		dev_err(wp->dev, "failed to map sync region\n");
+		return -EBUSY;
+	}
+
+	of_node_put(np);
+
+	return 0;
+}
+
+static int sprd_wcn_init_proc(struct sprd_wcn_shared *wcn,
+			      const struct sprd_wcn_subsys_info *info,
+			      const char *node_name,
+			      struct device *dev)
+{
+	struct sprd_wcn_proc *wp;
+	struct rproc *rproc;
+	const char *fw_name;
+	int ret;
+
+	dev->of_node = of_get_child_by_name(wcn->dev->of_node, node_name);
+	if (!dev->of_node) {
+		dev_err(wcn->dev, "unable to get %s child node\n", node_name);
+		return -ENOENT;
+	}
+
+	dev->parent = wcn->dev;
+	dev_set_name(dev, "%s:%s", dev_name(wcn->dev), node_name);
+
+	ret = device_register(dev);
+	if (ret < 0)
+		return ret;
+
+	ret = of_property_read_string_index(dev->of_node, "firmware-name",
+					    0, &fw_name);
+	if (ret < 0) {
+		dev_err(dev, "unable to read firmware-name\n");
+		goto err;
+	}
+
+	rproc = devm_rproc_alloc(dev, node_name, &wcn_ops,
+				 fw_name, sizeof(*wp));
+	if (!rproc)
+		goto err;
+
+	rproc->auto_boot = !strcmp(node_name, "btwf");
+
+	wp = rproc->priv;
+	wp->wcn = wcn;
+	wp->dev = dev;
+	wp->info = info;
+
+	ret = sprd_wcn_parse_memory_region(wp);
+	if (ret < 0)
+		goto err;
+
+	wp->wcn_ahb_regs = syscon_regmap_lookup_by_phandle(dev->of_node,
+							"sprd,syscon-wcn-ahb");
+	if (IS_ERR(wp->wcn_ahb_regs)) {
+		dev_err(wp->dev, "failed to get wcn-ahb syscon handle\n");
+		ret = PTR_ERR(wp->wcn_ahb_regs);
+		goto err;
+	}
+
+	ret = devm_sprd_rproc_add_sipc_subdev(rproc, dev->of_node, &wp->sipc);
+	if (ret < 0) {
+		dev_err(wp->dev, "failed to add SIPC subdev\n");
+		goto err;
+	}
+
+	ret = devm_rproc_add(wp->dev, rproc);
+	if (ret < 0) {
+		dev_err(wp->dev, "failed to register remoteproc\n");
+		goto err;
+	}
+
+	dev_set_drvdata(wp->dev, rproc);
+
+	return 0;
+
+err:
+	device_unregister(dev);
+	return ret;
+}
+
+static int sprd_wcn_probe(struct platform_device *pdev)
+{
+	struct sprd_wcn_shared *wcn;
+	int ret;
+
+	wcn = devm_kzalloc(&pdev->dev, sizeof(*wcn), GFP_KERNEL);
+	if (!wcn)
+		return -ENOMEM;
+
+	mutex_init(&wcn->lock);
+
+	wcn->dev = &pdev->dev;
+	platform_set_drvdata(pdev, wcn);
+
+	wcn->info = of_device_get_match_data(wcn->dev);
+	if (!wcn->info)
+		return -EINVAL;
+
+	ret = devm_pm_runtime_enable(wcn->dev);
+	if (ret < 0) {
+		dev_err(wcn->dev, "failed to enable runtime PM\n");
+		return ret;
+	}
+
+	ret = devm_regulator_bulk_get_const(wcn->dev, wcn->info->num_regulators,
+					    wcn->info->regulators, &wcn->regulators);
+	if (ret < 0) {
+		dev_err(wcn->dev, "failed to get regulators\n");
+		return ret;
+	}
+
+	wcn->en_gpio = devm_gpiod_get(wcn->dev, "enable", GPIOD_OUT_LOW);
+	if (IS_ERR(wcn->en_gpio)) {
+		dev_err(wcn->dev, "failed to get enable gpio\n");
+		return ret;
+	}
+
+	wcn->rst_gpio = devm_gpiod_get(wcn->dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(wcn->rst_gpio)) {
+		dev_err(wcn->dev, "failed to get reset gpio\n");
+		return ret;
+	}
+
+	wcn->ap_pub_apb_regs = syscon_regmap_lookup_by_phandle(wcn->dev->of_node,
+							       "sprd,syscon-ap-pub-apb");
+	if (IS_ERR(wcn->ap_pub_apb_regs)) {
+		dev_err(wcn->dev, "failed to get ap-pub-apb syscon handle\n");
+		return PTR_ERR(wcn->ap_pub_apb_regs);
+	}
+
+	wcn->ap_aon_apb_regs = syscon_regmap_lookup_by_phandle(wcn->dev->of_node,
+							       "sprd,syscon-ap-aon-apb");
+	if (IS_ERR(wcn->ap_aon_apb_regs)) {
+		dev_err(wcn->dev, "failed to get ap-aon-apb syscon handle\n");
+		return PTR_ERR(wcn->ap_aon_apb_regs);
+	}
+
+	wcn->wcn_aon_apb_regs = syscon_regmap_lookup_by_phandle(wcn->dev->of_node,
+								"sprd,syscon-wcn-aon-apb");
+	if (IS_ERR(wcn->wcn_aon_apb_regs)) {
+		dev_err(wcn->dev, "failed to get wcn-aon-apb syscon handle\n");
+		return PTR_ERR(wcn->wcn_aon_apb_regs);
+	}
+
+	wcn->wcn_aon_ahb_regs = syscon_regmap_lookup_by_phandle(wcn->dev->of_node,
+								"sprd,syscon-wcn-aon-ahb");
+	if (IS_ERR(wcn->wcn_aon_ahb_regs)) {
+		dev_err(wcn->dev, "failed to get wcn-aon-ahb syscon handle\n");
+		return PTR_ERR(wcn->wcn_aon_ahb_regs);
+	}
+
+	wcn->reset = devm_reset_control_get_shared(wcn->dev, NULL);
+	if (IS_ERR(wcn->reset)) {
+		dev_err(wcn->dev, "failed to get wcn reset\n");
+		return PTR_ERR(wcn->reset);
+	}
+
+	device_set_wakeup_capable(wcn->dev, true);
+
+	if (of_property_read_bool(wcn->dev->of_node, "wakeup-source"))
+		device_wakeup_enable(wcn->dev);
+
+	ret = sprd_wcn_init_proc(wcn, &wcn->info->btwf, "btwf", &wcn->btwf_dev);
+	if (ret < 0)
+		return ret;
+
+	ret = sprd_wcn_init_proc(wcn, &wcn->info->gnss, "gnss", &wcn->gnss_dev);
+	if (ret < 0) {
+		device_unregister(&wcn->btwf_dev);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void sprd_wcn_remove(struct platform_device *pdev)
+{
+	struct sprd_wcn_shared *wcn = platform_get_drvdata(pdev);
+
+	device_unregister(&wcn->btwf_dev);
+	device_unregister(&wcn->gnss_dev);
+}
+
+static const struct regulator_bulk_data ums9230_wcn_regulators[] = {
+	{ "vdd" },
+	{ "vddwifipa" },
+	{ "dcxo1v8" },
+};
+
+static const struct sprd_wcn_info ums9230_wcn_info = {
+	.regulators = ums9230_wcn_regulators,
+	.num_regulators = ARRAY_SIZE(ums9230_wcn_regulators),
+
+	.dram_offset = 0x87000000,
+
+	/* ap-pub-apb */
+	.mem_remap_reg = 0x9018,
+	.mem_remap_val = 0x70,
+
+	/* ap-aon-apb */
+	.ap_wcn_sleep_ctrl_reg = 0x034c,
+	.ap_wcn_sleep_allow_mask = BIT(16),
+
+	/* wcn-aon-apb */
+	.deep_sleep_ctrl_reg = 0x0090,
+	.deep_sleep_ctrl_mask = BIT(30),
+
+	.btwf = {
+		.init_status_offset = 0,
+		.init_status_boot_magic = 0xf0f0f0ff,
+
+		/* wcn-aon-ahb */
+		.sys_rst_reg = 0x000c,
+		.sys_rst_mask = BIT(6) | BIT(4) | BIT(2) | BIT(0),
+		.sync_addr_reg = 0x004c,
+
+		/* wcn-aon-apb */
+		.deep_sleep_reg = 0x0098,
+		.deep_sleep_mask = BIT(12) | BIT(3) | BIT(2),
+
+		/* wcn-btwf-ahb */
+		.boot_sel_reg = 0x0410,
+	},
+
+	.gnss = {
+		.init_status_offset = 0x34,
+		.init_status_boot_magic = 0x12345678,
+
+		/* wcn-aon-ahb */
+		.sys_rst_reg = 0x000c,
+		.sys_rst_mask = BIT(5) | BIT(1),
+		.sync_addr_reg = 0x0050,
+
+		/* wcn-aon-apb */
+		.deep_sleep_reg = 0x00c8,
+		.deep_sleep_mask = BIT(12) | BIT(3),
+
+		/* wcn-gnss-ahb */
+		.boot_sel_reg = 0x0404,
+	},
+};
+
+static const struct of_device_id sprd_wcn_of_match[] = {
+	{ .compatible = "sprd,ums9230-wcn", .data = &ums9230_wcn_info },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, sprd_wcn_of_match);
+
+static int sprd_wcn_suspend(struct device *dev)
+{
+	struct sprd_wcn_shared *wcn = dev_get_drvdata(dev);
+	struct rproc *proc;
+
+	if (device_may_wakeup(dev))
+		return 0;
+
+	proc = dev_get_drvdata(&wcn->btwf_dev);
+	if (proc->state == RPROC_RUNNING)
+		sprd_wcn_stop(proc);
+
+	proc = dev_get_drvdata(&wcn->gnss_dev);
+	if (proc->state == RPROC_RUNNING)
+		sprd_wcn_stop(proc);
+
+	return 0;
+}
+
+static int sprd_wcn_resume(struct device *dev)
+{
+	struct sprd_wcn_shared *wcn = dev_get_drvdata(dev);
+	struct rproc *proc;
+	int ret;
+
+	if (device_may_wakeup(dev))
+		return 0;
+
+	proc = dev_get_drvdata(&wcn->gnss_dev);
+	if (proc->state == RPROC_RUNNING) {
+		ret = sprd_wcn_start(proc);
+		if (ret < 0)
+			return ret;
+	}
+
+	proc = dev_get_drvdata(&wcn->btwf_dev);
+	if (proc->state == RPROC_RUNNING) {
+		ret = sprd_wcn_start(proc);
+		if (ret < 0)
+			goto stop_gnss;
+	}
+
+	return 0;
+
+stop_gnss:
+	proc = dev_get_drvdata(&wcn->gnss_dev);
+	if (proc->state == RPROC_RUNNING)
+		sprd_wcn_stop(proc);
+
+	return ret;
+}
+
+static struct dev_pm_ops sprd_wcn_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(sprd_wcn_suspend, sprd_wcn_resume)
+};
+
+static struct platform_driver sprd_wcn_driver = {
+	.probe = sprd_wcn_probe,
+	.remove = sprd_wcn_remove,
+	.driver = {
+		.name = "sprd-wcn",
+		.of_match_table = sprd_wcn_of_match,
+		.pm = &sprd_wcn_pm_ops,
+	},
+};
+
+module_platform_driver(sprd_wcn_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_DESCRIPTION("Unisoc WCN remoteproc driver");
